### PR TITLE
[Merged by Bors] - feat: port Algebra.Hom.Equiv.Basic

### PR DESCRIPTION
**Pull request by winstonyin**
mathlib3 SHA: 76171581280d5b5d1e2d1f4f37e5420357bdc636

### CODE REVIEW

File: Mathlib/Algebra/Hom/Equiv/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,659 @@
...
+#align mul_equiv.coe_to_mul_hom MulEquiv.coe_toMulHom
+@[simp]
+theorem _root_.AddEquiv.coe_toAddHom [Add M] [Add N] {f : M ≃+ N} : f.toAddHom = f.toFun :=
+  rfl
+#align add_equiv.coe_to_add_hom AddEquiv.coe_toAddHom
+
+/-- A multiplicative isomorphism preserves multiplication. -/
+@[to_additive "An additive isomorphism preserves addition."]
+protected theorem map_mul (f : M ≃* N) : ∀ x y, f (x * y) = f x * f y :=
+  _root_.map_mul f
```
Comment (winstonyin) [inline Mathlib/Algebra/Hom/Equiv/Basic.lean:199]:
```
failed to synthesize
  MulHomClass (M ≃* N) ?m.239836 ?m.239837
(deterministic) timeout at 'typeclass', maximum number of heartbeats (20000) has been reached
```

**Comment by kbuzzard** [comment]
On an unrelated note, was the removal of all the blank lines in `Algebra.Hom.Group` intentional?

**Comment by winstonyin** [comment]
> On an unrelated note, was the removal of all the blank lines in `Algebra.Hom.Group` intentional?

Huh that's strange. Blank lines restored.

File: Mathlib/Algebra/Hom/Equiv/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,659 @@
...
+def refl (M : Type _) [Mul M] : M ≃* M :=
+  { Equiv.refl _ with map_mul' := fun _ _ => rfl }
+#align mul_equiv.refl MulEquiv.refl
+
+@[to_additive]
+instance : Inhabited (M ≃* M) :=
+  ⟨refl M⟩
+
+/-- The inverse of an isomorphism is an isomorphism. -/
+@[symm, to_additive "The inverse of an isomorphism is an isomorphism."]
```
Comment (winstonyin) [inline Mathlib/Algebra/Hom/Equiv/Basic.lean:234]: Bug with `symm` it seems [#855](https://github.com/leanprover-community/mathlib4/pull/855)

File: Mathlib/Algebra/Hom/Equiv/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,659 @@
...
+  rfl
+#align mul_equiv.symm_mk MulEquiv.symm_mk
+
+@[simp, to_additive]
+theorem refl_symm : (refl M).symm = refl M :=
+  rfl
+#align mul_equiv.refl_symm MulEquiv.refl_symm
+
+/-- Transitivity of multiplication-preserving isomorphisms -/
+@[trans, to_additive "Transitivity of addition-preserving isomorphisms"]
```
Comment (winstonyin) [inline Mathlib/Algebra/Hom/Equiv/Basic.lean:293]: Bug with `trans`, similar to `symm` [#855](https://github.com/leanprover-community/mathlib4/pull/855)

**Comment by semorrison** [comment]
Okay, we finally [sorted this out](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/typeclass.20timeout/near/314173235) thanks to a new linter (and Gabriel!)

**Comment by semorrison** [comment]
Okay, there are now lots of linting failures to deal with.

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**